queries/blockwise-discount-factors.sql

```diff
@@ -0,0 +1,129 @@
+with 
+sourced_cow_transfers as (
+    -- Incoming COW
+    select evt_tx_hash,
+           evt_block_time,
+           evt_block_number,
+           "to" as account,
+           value as cow_amount,
+           0 as vcow_amount
+    from cow_protocol."CowProtocolToken_evt_Transfer"
+    union all
+    -- Outgoing COW
+    select evt_tx_hash,
+           evt_block_time,
+           evt_block_number,
+           "from"     as account,
+           -1 * value as cow_amount,
+           0 as vcow_amount
+    from cow_protocol."CowProtocolToken_evt_Transfer"
+    union all
+    -- Incoming vCOW
+    select evt_tx_hash,
+           evt_block_time,
+           evt_block_number,
+           "to" as account,
+           0 as cow_amount,
+           value as vcow_amount
+    from cow_protocol."CowProtocolVirtualToken_evt_Transfer"
+    union
+    -- Outgoing vCOW
+    select evt_tx_hash,
+           evt_block_time,
+           evt_block_number,
+           "from"     as account,
+           0 as cow_amount,
+           -1 * value as vcow_amount
+    from cow_protocol."CowProtocolVirtualToken_evt_Transfer"
+),
+
+balances_at_block as (
+    select 
+        evt_block_time,
+        evt_block_number,
+        account,
+        cow_amount as cow_delta,
+        sum(cow_amount) OVER (PARTITION BY account ORDER BY evt_block_number) as cow_balance,
+        vcow_amount as vcow_delta,
+        sum(vcow_amount) OVER (PARTITION BY account ORDER BY evt_block_number) as vcow_balance
+    from sourced_cow_transfers
+),
+
+blockwise_combined_balances as (
+    select
+        evt_block_time,
+        evt_block_number,
+        account,
+        (vcow_balance + cow_balance) / 10 ^18 as combined_balance
+    from balances_at_block
+    where evt_block_number > 14475154
+    and (vcow_balance + cow_balance) > 10^21
+    and account != '\x0000000000000000000000000000000000000000'
+),
+-- Snapshot
+-- https://snapshot.org/#/cow.eth/proposal/0x4bb9b614bdc4354856c4d0002ad0845b73b5290e5799013192cbc6491e6eea0e
+-- Tier 1 1,000 COW Discount: 5%
+-- Tier 2 10,000 COW Discount: 10%
+-- Tier 3 100,000 COW Discount: 20%
+-- Tier 4 1,000,000 COW Discount: 40%
+subsidy_tiers as (
+    SELECT * FROM (VALUES
+        -- (0, 10^3, 0.0),
+        (10^3, 10^4, 0.05),
+        (10^4, 10^5, 0.1),
+        (10^5, 10^6, 0.2),
+        (10^6, 10^12, 0.4) -- use an essentially infinite upper bound for the top tier!
+    ) as t (lower_bound, upper_bound, discount)
+),
+
+blockwise_discount as (
+    select
+        evt_block_time,
+        evt_block_number,
+        account,
+        combined_balance,
+        (select max(discount) from subsidy_tiers where combined_balance >= lower_bound) as discount
+    from blockwise_combined_balances
+),
+
+fee_discounts as (
+    select
+        trader,
+        block_time,
+        max(evt_block_time) balance_from,
+        fee_usd as fee_paid,
+        discount
+    from gnosis_protocol_v2."trades" t
+    join blockwise_discount
```

Comment:
You do not specify which join it is. I think its nice to make it explicity. 

I don't understand why the query works. :( since you are grouping by discount, you should get  - according to my understanding - several rows for 1 trade with the different discounts, with all the different discount rates a trader ever had in my mind. 
But it seems that I am missing something. Could you tell me why it works?

Reply:
join is by default an inner join. I can make it explicit if you want but this is generally considered a standard in all sql dialects.

This is an optimization technique that was required for this query not to timeout. The idea for trick here is the choice of max block time however I think you suspicions are correct and we definitely need to check on this....

I would very much appreciate your reply to this [slack discussion](https://cowservices.slack.com/archives/C0361CCTFD5/p1655452963173299) I have been having with myself.

The real goal of the problem is to avoid using the natural sub query approach:

```sql
select 
    trader,
    block_time,
    (
        select balance 
        from balances b
        where t.trader = b.account 
        and b.block_time < t.block_time
        order by b.block_number desc
        limit 1
    ) as balance
from trades t
```

Reply:
Actually, it looks like you have participated a very little bit already in the discussion:

https://cowservices.slack.com/archives/C0361CCTFD5/p1655454201924509?thread_ts=1655452963.173299&cid=C0361CCTFD5

Maybe you want to double check.

Reply:
Cool. I posted into the slack a potential solution

Reply:
Yes, this looks good, but I think the ordering clause will select the earliest block time satisfying `b.block_time < block_time` where we want the latest. I like it and will give it a try.

Reply:
Ok so far this is not working at all, in fact the proposed solution is also one resulting in multiple rows per trade for every balance update on the account.

I will continue to play with this, but am not having any luck at the moment.

Reply:
Ok! I have something that is working.

```
fee_discounts as (
    select distinct on (order_uid)
        trader,
        block_time,
        evt_block_time,
        fee_usd as fee_paid,
        discount
    from gnosis_protocol_v2."trades" t
    inner join blockwise_discount
        on account = trader
        and evt_block_time < block_time
    -- This is the transaction where we got the block time and block number being used immediately below. 
    -- Specifically the transaction where the 60 ETH was transferred.
    -- https://etherscan.io/tx/0x1100cd4a50a2c224ec39f861aef7574df394edb2b5ed850705cf0bb34f6a300d
    where block_time > '2022-03-28 14:19' -- number = 14475154
    order by order_uid, evt_block_time desc
),
```

We only needed to be distinct on order uid. Note that this might not be accurate for partially fillable orders, but I think this is pretty good for now! 

Turns out we have only spent 30 ETH so far!

Reply:
To verify this look at this wallet. 

Who has made a trade and has several balance changes before and after the trade

`\x3a11f4c84688a1264690d696d8d807a25ee02dd2`